File: leetcode/0155(最小栈).cpp
#include <iostream>
#include <stack>
#include <queue>
#include <list>

using namespace std;

class MinStack {
public:
    /** initialize your data structure here. */
    list<int> L;
    MinStack() {

    }

    void push(int x) {
        L.push_front(x);
    }

    void pop() {
        L.pop_front();
    }

    int top() {
        return *L.begin();
    }

    int getMin() {
        int _min = *L.begin();
        auto it = L.begin();
        ++it;
        while(it!=L.end()){
            if(*it < _min) _min = *it;
            ++it;
        }
        return _min;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */


int main(){


    cout<<pq.top()<<endl;

    return 0;
}

File: leetcode/0155(最小栈)/0155(最小栈).go
type MinStack struct {
    data, min []int 
}


func Constructor() MinStack {
    return MinStack{data: []int{}, min: []int{}}
}


func (this *MinStack) Push(val int)  {
    if len(this.data) > 0 {
        if val <= this.min[len(this.min)-1] {
            this.min = append(this.min, val)
        }
        this.data = append(this.data, val)
    } else {
        this.min = append(this.min, val)
        this.data = append(this.data, val)
    }
    
    
}


func (this *MinStack) Pop()  {
    if len(this.data) > 0 {
        if this.data[len(this.data)-1] == this.min[len(this.min)-1] {
            this.min = this.min[:len(this.min)-1]
        }
        this.data = this.data[:len(this.data)-1]
    }
}


func (this *MinStack) Top() int {
    return this.data[len(this.data)-1]
}


func (this *MinStack) GetMin() int {
    return this.min[len(this.min)-1]
}